stdlib/random.gr

```diff
@@ -0,0 +1,176 @@
+import WasiRandom from "sys/random"
+import Result from "result"
+import Int32 from "int32"
+import Int64 from "int64"
+import WasmI32 from "runtime/unsafe/wasmi32"
+import WasmI64 from "runtime/unsafe/wasmi64"
+import Memory from "runtime/unsafe/memory"
+import DS from "runtime/dataStructures"
+
+record Random {
+  seed: Int64,
+  mut counter: Int64,
+  mut initialized: Bool,
+}
+
+let incCounter = random => {
+  random.counter = Int64.incr(random.counter)
+}
+
+// https://arxiv.org/pdf/2004.06278v3.pdf
+@disableGC
+let rec squares = (ctr: Int64, key: Int64) => {
+  // Need to implement with @disableGC because
+  // overflows are part of this algorithm
+  let ctrArg = ctr
+  let keyArg = key
+  let ctr = DS.loadInt64(WasmI32.fromGrain(ctr))
+  let key = DS.loadInt64(WasmI32.fromGrain(key))
+  let mut x = WasmI64.mul(ctr, key)
+  let mut y = x
+  let mut z = WasmI64.add(y, key)
+  // round 1
+  x = WasmI64.add(WasmI64.mul(x, x), y)
+  x = WasmI64.or(WasmI64.shrU(x, 32N), WasmI64.shl(x, 32N))
+  // round 2
+  x = WasmI64.add(WasmI64.mul(x, x), z)
+  x = WasmI64.or(WasmI64.shrU(x, 32N), WasmI64.shl(x, 32N))
+  // round 3
+  x = WasmI64.add(WasmI64.mul(x, x), y)
+  x = WasmI64.or(WasmI64.shrU(x, 32N), WasmI64.shl(x, 32N))
+  let ret = WasmI32.wrapI64(
+    WasmI64.shrU(WasmI64.add(WasmI64.mul(x, x), z), 32N)
+  )
+  let ret = WasmI32.toGrain(DS.newInt32(ret)): Int32
+  Memory.decRef(WasmI32.fromGrain(ctrArg))
+  Memory.decRef(WasmI32.fromGrain(keyArg))
+  Memory.decRef(WasmI32.fromGrain(squares))
+  ret
+}
+
+// [NOTE] The PRNG algorithm seems to not work as well with small seeds, but
+// users are more likely to use small numbers for their seeds. Consequently,
+// we produce an actual seed xoring feeding the given seed with a
+// constant (large) number.
+
+/**
+ * Creates a new random number generator with the given seed.
+ *
+ * @param seed: The seed for the random number generator
+ * @returns The random number generator
+ */
+export let make = seed => {
+  { seed, counter: 0L, initialized: false }
+}
+
+/**
+ * Creates a new random number generator with a random seed.
+ *
+ * @param seed: The seed for the random number generator
+ * @returns The random number generator
+ */
+export let makeUnseeded = () => {
+  // TODO: Should we just .expect this result for UX's sake?
```

Comment:
This is hard to say. Personally, I think it's "easier" for us to force the `.expect` calls onto the user. Then, they can also audit their code for crashes more easily.

Reply:
@peblair @ospencer did you ever decide on what you wanted? Should we just open an issue to evaluate/re-evaluate?

Reply:
I could go either way. I don't really want to make an issue since I feel like it'll kinda just sit around. For that reason I lean more towards doing the "correct" thing by just handing over a result and letting them decide. That way just avoids any problem when someone actually cares about what happens from WASI.